File: data/src/main/kotlin/dev/yashgarg/kimai/util/ApiResult.kt
/*
 * Copyright © 2023 Yash Garg.
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */
package dev.yashgarg.kimai.util

import retrofit2.HttpException
import retrofit2.Response

sealed interface ApiResult<T : Any>

class ApiSuccess<T : Any>(val data: T) : ApiResult<T>

class ApiError<T : Any>(val code: Int, val message: String?) : ApiResult<T>

class ApiException<T : Any>(val e: Throwable) : ApiResult<T>

suspend fun <T : Any> handleApi(execute: suspend () -> Response<T>): ApiResult<T> {
  return try {
    val response = execute()
    val body = response.body()

    if (response.isSuccessful && body != null) {
      ApiSuccess(body)
    } else {
      ApiError(code = response.code(), message = response.message())
    }
  } catch (e: HttpException) {
    ApiError(code = e.code(), message = e.message())
  } catch (e: Throwable) {
    ApiException(e)
  }
}

File: lib/utils/api_result.dart
import 'dart:developer';

import 'package:dio/dio.dart';

sealed class ApiResult<dT> {}

class ApiSuccess<T> extends ApiResult<T> {
  final T data;

  ApiSuccess(this.data);
}

class ApiError<T> extends ApiResult<T> {
  final int code;
  final String? message;

  ApiError({
    required this.code,
    required this.message,
  });
}

Future<ApiResult<T>> handleApi<T>(
  Future<Response<T>> Function() execute,
) async {
  try {
    final response = await execute();

    log('Response: ${response.data}');

    // ignore: null_check_on_nullable_type_parameter
    return ApiSuccess(response.data!);
  } on DioException catch (e) {
    log('DioException: ${e.message}');

    if (e.response != null) {
      return ApiError(
        code: e.response!.statusCode!,
        message: e.response!.statusMessage,
      );
    } else {
      return ApiError(
        code: -1,
        message: e.message,
      );
    }
  } on Exception catch (e) {
    log('Exception: ${e.toString()}');

    return ApiError(
      code: -1,
      message: e.toString(),
    );
  }
}
